Bug Report: compiled EvalEngine corrupts the VM stack when a later LOCATE or SUBSTRING operand is NULL

### Overview of the Issue

The compiled EvalEngine leaves stale operands on the VM stack when a later argument of `LOCATE` or three-argument `SUBSTRING` evaluates to `NULL`. The null-check emitted for those arguments jumps past the consuming instruction without collapsing the operands that are already on the stack. An enclosing expression then either consumes a leftover operand and returns an incorrect result, or pushes past the precomputed stack size and panics with `index out of range` inside `EvaluateVM`, which has no panic recovery.

This is the same defect class as #20623, which was fixed for `LIKE` in #20624. Auditing the remaining null-check call sites in the compiler turned up two more affected builtins:

- `builtinLocate.compile` checks its second and third operands with `compileNullCheck1` (`go/vt/vtgate/evalengine/fn_string.go:1699` and `:1727`), which pops nothing before jumping.
- `builtinSubstring.compile` checks a `NULL` length with `compileNullCheck2(str, l)` (`go/vt/vtgate/evalengine/fn_string.go:1630`), which collapses only two of the three slots on the stack, leaving the string operand behind.

All other multi-operand `compile` implementations follow the correct pattern (`compileNullCheck1` for the first operand, `compileNullCheck1r` for the second, `compileNullCheck3`/`compileNullCheckArg` for later ones, or purely local `compileNullCheckOffset` jumps that preserve stack balance).

Because constant folding evaluates all-literal expressions through the (correct) interpreted path, the bug only triggers when the `NULL` arrives dynamically, e.g. from a nullable column or a bind variable.

### Reproduction Steps

Evaluating the following expressions with `column0 = NULL` through the compiled EvalEngine (interpreted evaluation and MySQL both handle them correctly):

```sql
NULL <=> LOCATE('foo', column0)         -- returns 0, expected 1
IFNULL(LOCATE('foo', column0), 7)       -- panics: index out of range in push_null
NULL <=> SUBSTRING('abc', 1, column0)   -- returns 0, expected 1
IFNULL(SUBSTRING('abc', 1, column0), 'x') -- panics: index out of range in push_null
```

Panic from the `LOCATE` reproduction:

```
panic: runtime error: index out of range [4] with length 4

vitess.io/vitess/go/vt/vtgate/evalengine.push_null(...)
	go/vt/vtgate/evalengine/compiler_asm_push.go:31
vitess.io/vitess/go/vt/vtgate/evalengine.(*ExpressionEnv).EvaluateVM(...)
	go/vt/vtgate/evalengine/vm.go:85
```

A self-contained regression harness (mirroring `TestCompilerSingle`) that reproduces all four cases:

```go
venv := vtenv.NewTestEnv()
expr, _ := venv.Parser().ParseExpr(`ifnull(locate('foo', column0), 7)`)
fields := evalengine.FieldResolver(makeFields([]sqltypes.Value{sqltypes.NULL}))
cfg := &evalengine.Config{
	ResolveColumn: fields.Column,
	ResolveType:   fields.Type,
	Collation:     collations.CollationUtf8mb4ID,
	Environment:   venv,
}
converted, _ := evalengine.Translate(expr, cfg) // constant folding enabled, as in production
env := evalengine.NewExpressionEnv(context.Background(), nil, evalengine.NewEmptyVCursor(venv, time.UTC))
env.Row = []sqltypes.Value{sqltypes.NULL}
env.Evaluate(converted) // panics
```

### Binary Version

Reproduced on current `main` (e15cc4c730 and earlier). The affected code is also present in the supported release branches (verified on v23.0.5 and v24.0.2).

### Operating System and Environment details

Linux arm64 (development environment); the defect is platform-independent.

### Log Fragments

See panic trace above.

--- a/go/vt/vtgate/evalengine/compiler_test.go
+++ b/go/vt/vtgate/evalengine/compiler_test.go
@@ -202,6 +202,36 @@ func TestCompilerSingle(t *testing.T) {
 			result:     "INT64(2)",
 		},
 		{
+			// A NULL string operand must collapse both LOCATE operands before outer
+			// expressions consume the result.
+			expression: "null <=> locate('foo', column0)",
+			values:     []sqltypes.Value{sqltypes.NULL},
+			result:     "INT64(1)",
+		},
+		{
+			// A leftover operand would make IFNULL overflow the VM stack.
+			expression: "ifnull(locate('foo', column0), 7)",
+			values:     []sqltypes.Value{sqltypes.NULL},
+			result:     "INT64(7)",
+		},
+		{
+			// A NULL position must collapse all three LOCATE operands.
+			expression: "ifnull(locate('foo', 'foobar', column0), 7)",
+			values:     []sqltypes.Value{sqltypes.NULL},
+			result:     "INT64(7)",
+		},
+		{
+			// A NULL length must collapse all three SUBSTRING operands.
+			expression: "null <=> substring('abc', 1, column0)",
+			values:     []sqltypes.Value{sqltypes.NULL},
+			result:     "INT64(1)",
+		},
+		{
+			expression: "ifnull(substring('abc', 1, column0), 'x')",
+			values:     []sqltypes.Value{sqltypes.NULL},
+			result:     `VARCHAR("x")`,
+		},
+		{
 			expression: "1 + column0",
 			values:     []sqltypes.Value{sqltypes.NewFloat64(1)},
 			result:     "FLOAT64(2)",

--- a/go/vt/vtgate/evalengine/fn_string.go
+++ b/go/vt/vtgate/evalengine/fn_string.go
@@ -1627,7 +1627,7 @@ func (call *builtinSubstring) compile(c *compiler) (ctype, error) {
 		if err != nil {
 			return ctype{}, err
 		}
-		skip2 = c.compileNullCheck2(str, l)
+		skip2 = c.compileNullCheckArg(l, 2)
 		_ = c.compileToInt64(l, 1)
 		c.asm.Fn_SUBSTRING3(tt, cs, col)
 	} else {
@@ -1696,7 +1696,7 @@ func (call *builtinLocate) compile(c *compiler) (ctype, error) {
 		return ctype{}, err
 	}
 
-	skip2 := c.compileNullCheck1(str)
+	skip2 := c.compileNullCheck1r(str)
 	var skip3 *jump
 
 	if !str.isTextual() {
@@ -1724,7 +1724,7 @@ func (call *builtinLocate) compile(c *compiler) (ctype, error) {
 		if err != nil {
 			return ctype{}, err
 		}
-		skip3 = c.compileNullCheck1(l)
+		skip3 = c.compileNullCheckArg(l, 2)
 		_ = c.compileToInt64(l, 1)
 	}
 

